SimPEG/PF/Magnetics.py

```diff
@@ -23,361 +26,476 @@ class MagneticIntegral(Problem.LinearProblem):
+        """
+            Call for general mapping of the problem
+        """
+        return self.chiMap
 
-            # Loop through all observations and create forward operator (nD-by-nC)
-            print("Begin calculation of forward operator: " + Magnetization)
+    @property
+    def dfdm(self):
 
-        # Add counter to dsiplay progress. Good for large problems
-        count = -1
-        for ii in range(ndata):
+        if self.model is None:
+            raise Exception('Problem needs a chi chi')
 
-            tx, ty, tz = get_T_mat(Xn, Yn, Zn, rxLoc[ii, :])
+        if getattr(self, '_dfdm', None) is None:
 
-            if self.forwardOnly:
+            Bxyz = self.Bxyz_a(self.chiMap * self.model)
 
-                if self.rtype == 'tmi':
-                    fwr_out[ii] = (Ptmi.dot(np.vstack((tx, ty, tz)))*Mxyz).dot(m)
+            Bx = sp.spdiags(Bxyz[:, 0], 0, self.nD, self.nD)
+            By = sp.spdiags(Bxyz[:, 1], 0, self.nD, self.nD)
+            Bz = sp.spdiags(Bxyz[:, 2], 0, self.nD, self.nD)
 
-                elif self.rtype == 'xyz':
-                    fwr_out[ii] = (tx*Mxyz).dot(m)
-                    fwr_out[ii+ndata] = (ty*Mxyz).dot(m)
-                    fwr_out[ii+2*ndata] = (tz*Mxyz).dot(m)
+            self._dfdm = sp.hstack((Bx, By, Bz))
 
-            else:
+        return self._dfdm
 
-                if Magnetization == 'ind':
+    def Bxyz_a(self, m):
+        """
+            Return the normalized B fields
+        """
 
-                    if survey.srcField.rxList[0].rxType == 'tmi':
-                        fwr_out[ii, :] = Ptmi.dot(np.vstack((tx, ty, tz)))*Mxyz
+        # Get field data
+        if self.coordinate_system == 'spherical':
+            m = matutils.atp2xyz(m)
 
-                    elif survey.srcField.rxList[0].rxType == 'xyz':
-                        fwr_out[ii, :] = tx*Mxyz
-                        fwr_out[ii+ndata, :] = ty*Mxyz
-                        fwr_out[ii+2*ndata, :] = tz*Mxyz
+        if getattr(self, '_Mxyz', None) is not None:
+            Bxyz = np.dot(self.G, (self.Mxyz*m).astype(np.float32))
+        else:
+            Bxyz = np.dot(self.G, m.astype(np.float32))
 
-                elif Magnetization == 'xyz':
+        amp = self.calcAmpData(Bxyz.astype(np.float64))
+        Bamp = sp.spdiags(1./amp, 0, self.nD, self.nD)
 
-                    if survey.srcField.rxList[0].rxType == 'tmi':
-                        fwr_out[ii, :] = Ptmi.dot(np.vstack((tx, ty, tz)) *
-                                                  survey.srcField.param[0])
+        return Bamp*Bxyz.reshape((self.nD, 3), order='F')
 
-                    elif survey.srcField.rxList[0].rxType == 'xyz':
-                        fwr_out[ii, :] = tx * survey.srcField.param[0]
-                        fwr_out[ii+ndata, :] = ty * survey.srcField.param[0]
-                        fwr_out[ii+2*ndata, :] = tz * survey.srcField.param[0]
+    def Intrgl_Fwr_Op(self, m=None, magType='H0', rxType='tmi'):
```

Comment:
Does `integral_forward_operator` work?

also, what is `magType`? is there something more descriptive we can use?